File: backend/app/core/compliance_rules.py
"""
Deterministic compliance rule engine — no LLM needed.
Rules based on RBI IRAC norms, PSL guidelines, PMLA thresholds, and DPDP Act.
"""
from typing import List, Dict, Any

# PII field name patterns (case-insensitive substring match)
PII_PATTERNS = [
    "aadhaar", "aadhar", "uid", "biometric", "fingerprint", "iris",
    "pan_number", "pan number", "passport", "voter_id", "voter id",
    "ssn", "social security", "dob", "date_of_birth", "date of birth",
    "password", "pin", "cvv", "credit_card", "debit_card",
]


def check_pii_query(question: str) -> Dict | None:
    """
    Pre-screen: block queries that try to expose PII.
    Returns a block dict if triggered, None if clean.
    """
    q = question.lower()
    for pattern in PII_PATTERNS:
        if pattern in q:
            return {
                "rule": "PII_EXPOSURE",
                "status": "blocked",
                "message": (
                    f"Query blocked: This analysis appears to request access to sensitive "
                    f"personal data ('{pattern}'). Per the DPDP Act 2023, PII including Aadhaar, "
                    f"PAN, biometrics, and financial credentials must be anonymised before analysis. "
                    f"Use the Sensitive Column toggle to mask this field first."
                ),
            }
    return None


def check_npa_classification(data: List[Dict], days_col: str = None, class_col: str = None) -> Dict | None:
    """
    Post-validate: check for loans marked Standard with DPD in SMA-2 range (61–90 days).
    Returns annotation dict or None if clean.
    """
    if not data:
        return None

    # Auto-detect columns
    if not days_col or not class_col:
        sample = data[0]
        keys = list(sample.keys())
        for k in keys:
            kl = k.lower()
            if days_col is None and any(x in kl for x in ["days_past", "dpd", "overdue", "past_due"]):
                days_col = k
            if class_col is None and any(x in kl for x in ["classification", "npa", "asset_class", "status"]):
                class_col = k

    if not days_col or not class_col:
        return None

    misclassified = []
    for row in data:
        try:
            dpd = int(row.get(days_col, 0) or 0)
            cls = str(row.get(class_col, "")).strip()
            if 61 <= dpd <= 90 and cls.lower() in ("standard", "regular"):
                misclassified.append(row)
        except (ValueError, TypeError):
            continue

    if not misclassified:
        return {
            "rule": "NPA_CLASSIFICATION",
            "status": "compliant",
            "message": "NPA classification verified. All accounts with DPD 61–90 days are correctly marked SMA-2 per RBI IRAC norms.",
        }

    count = len(misclassified)
    return {
        "rule": "NPA_CLASSIFICATION",
        "status": "warning",
        "message": (
            f"{count} loan(s) appear misclassified: they have DPD between 61–90 days but are "
            f"still marked 'Standard'. Per RBI IRAC norms, these should be classified as SMA-2. "
            f"Misclassification may lead to regulatory findings and provisioning shortfalls."
        ),
        "affected_count": count,
    }


def check_psl_ratio(data: List[Dict], sector_col: str = None, amount_col: str = None) -> Dict | None:
    """
    Post-validate: calculate PSL ratio and flag if below 40% RBI target.
    """
    if not data:
        return None

    PSL_SECTORS = {"agriculture", "msme", "education", "housing", "weaker sections", "priority sector"}

    if not sector_col or not amount_col:
        sample = data[0]
        keys = list(sample.keys())
        for k in keys:
            kl = k.lower()
            if sector_col is None and any(x in kl for x in ["sector", "category", "purpose", "segment"]):
                sector_col = k
            if amount_col is None and any(x in kl for x in ["amount", "principal", "outstanding", "balance", "loan_amount"]):
                amount_col = k

    if not sector_col or not amount_col:
        return None

    total_amt = 0.0
    psl_amt = 0.0
    for row in data:
        try:
            amt = float(row.get(amount_col, 0) or 0)
            sector = str(row.get(sector_col, "")).strip().lower()
            total_amt += amt
            if any(ps in sector for ps in PSL_SECTORS):
                psl_amt += amt
        except (ValueError, TypeError):
            continue

    if total_amt == 0:
        return None

    ratio = psl_amt / total_amt * 100
    target = 40.0
    shortfall_pct = target - ratio

    if ratio >= target:
        return {
            "rule": "PSL_RATIO",
            "status": "compliant",
            "message": f"PSL ratio is {ratio:.1f}%, meeting the RBI mandate of 40%. No shortfall.",
        }

    shortfall_amt = (target / 100 - psl_amt / total_amt) * total_amt
    return {
        "rule": "PSL_RATIO",
        "status": "warning",
        "message": (
            f"PSL ratio is {ratio:.1f}%, below the RBI mandate of 40%. "
            f"Shortfall: {shortfall_pct:.1f} percentage points "
            f"(approximately \u20b9{shortfall_amt/1e7:.1f} Cr). "
            f"Failure to meet PSL targets results in deposits in RIDF at below-market rates."
        ),
        "psl_ratio": round(ratio, 2),
        "target": target,
    }


def check_pmla_threshold(data: List[Dict], amount_col: str = None, mode_col: str = None) -> Dict | None:
    """
    Post-validate: flag cash transactions above Rs.10 lakh (CTR threshold).
    """
    if not data:
        return None

    THRESHOLD = 1_000_000  # Rs.10 lakh

    if not amount_col or not mode_col:
        sample = data[0]
        keys = list(sample.keys())
        for k in keys:
            kl = k.lower()
            if amount_col is None and any(x in kl for x in ["amount", "value", "txn_amount"]):
                amount_col = k
            if mode_col is None and any(x in kl for x in ["mode", "channel", "type", "method"]):
                mode_col = k

    if not amount_col:
        return None

    triggers = []
    for row in data:
        try:
            amt = float(row.get(amount_col, 0) or 0)
            mode = str(row.get(mode_col, "") or "").lower() if mode_col else ""
            if amt >= THRESHOLD and ("cash" in mode or mode == ""):
                triggers.append(row)
        except (ValueError, TypeError):
            continue

    if not triggers:
        return None

    count = len(triggers)
    return {
        "rule": "PMLA_CTR_THRESHOLD",
        "status": "warning",
        "message": (
            f"{count} transaction(s) meet the PMLA Cash Transaction Report (CTR) threshold "
            f"of \u20b910 lakh. Per PMLA 2002, CTRs must be filed with FIU-IND by the 15th of "
            f"the following month. Verify that CTRs have been submitted for these transactions."
        ),
        "affected_count": count,
    }


def run_all_applicable_rules(question: str, data: List[Dict]) -> List[Dict]:
    """
    Run all deterministic rules that are applicable based on question keywords and data.
    Returns list of annotation dicts.
    """
    annotations = []
    q = question.lower()

    # NPA / asset classification check
    if any(kw in q for kw in ["npa", "asset class", "classification", "overdue", "past due", "dpd", "sma", "standard"]):
        result = check_npa_classification(data)
        if result:
            annotations.append(result)

    # PSL check
    if any(kw in q for kw in ["psl", "priority sector", "agriculture", "msme", "education", "housing", "weaker"]):
        result = check_psl_ratio(data)
        if result:
            annotations.append(result)

    # PMLA / cash transaction check
    if any(kw in q for kw in ["cash", "transaction", "ctر", "pmla", "aml", "suspicious", "structuring", "10 lakh", "threshold"]):
        result = check_pmla_threshold(data)
        if result:
            annotations.append(result)

    return annotations

File: backend-node/src/core/complianceRules.ts
/**
 * Deterministic compliance rule engine — mirrors Python backend/app/core/compliance_rules.py.
 */
import { ParsedRow } from '../types';

const PII_PATTERNS = [
  'aadhaar','aadhar','uid','biometric','fingerprint','iris',
  'pan_number','pan number','passport','voter_id','voter id',
  'ssn','social security','dob','date_of_birth','date of birth',
  'password','pin','cvv','credit_card','debit_card',
];

export function checkPiiQuery(question: string): Record<string, string> | null {
  const q = question.toLowerCase();
  for (const pattern of PII_PATTERNS) {
    if (q.includes(pattern)) {
      return {
        rule: 'PII_EXPOSURE',
        status: 'blocked',
        message: `Query blocked: This analysis appears to request access to sensitive personal data ('${pattern}'). Per the DPDP Act 2023, PII including Aadhaar, PAN, biometrics, and financial credentials must be anonymised before analysis. Use the Sensitive Column toggle to mask this field first.`,
      };
    }
  }
  return null;
}

function checkNpaClassification(data: ParsedRow[]): Record<string, any> | null {
  if (!data.length) return null;
  const keys = Object.keys(data[0]);
  let daysCol: string | null = null;
  let classCol: string | null = null;
  for (const k of keys) {
    const kl = k.toLowerCase();
    if (!daysCol && ['days_past','dpd','overdue','past_due'].some((x) => kl.includes(x))) daysCol = k;
    if (!classCol && ['classification','npa','asset_class','status'].some((x) => kl.includes(x))) classCol = k;
  }
  if (!daysCol || !classCol) return null;
  const misclassified = data.filter((row) => {
    const dpd = parseInt(String(row[daysCol!] ?? 0), 10);
    const cls = String(row[classCol!] ?? '').trim().toLowerCase();
    return dpd >= 61 && dpd <= 90 && ['standard', 'regular'].includes(cls);
  });
  if (!misclassified.length) {
    return { rule: 'NPA_CLASSIFICATION', status: 'compliant', message: 'NPA classification verified. All accounts with DPD 61–90 days are correctly marked SMA-2 per RBI IRAC norms.' };
  }
  return {
    rule: 'NPA_CLASSIFICATION', status: 'warning',
    message: `${misclassified.length} loan(s) appear misclassified: DPD between 61–90 days but still marked 'Standard'. Per RBI IRAC norms, these should be SMA-2.`,
    affected_count: misclassified.length,
  };
}

const PSL_SECTORS = new Set(['agriculture','msme','education','housing','weaker sections','priority sector']);

function checkPslRatio(data: ParsedRow[]): Record<string, any> | null {
  if (!data.length) return null;
  const keys = Object.keys(data[0]);
  let sectorCol: string | null = null;
  let amountCol: string | null = null;
  for (const k of keys) {
    const kl = k.toLowerCase();
    if (!sectorCol && ['sector','category','purpose','segment'].some((x) => kl.includes(x))) sectorCol = k;
    if (!amountCol && ['amount','principal','outstanding','balance','loan_amount'].some((x) => kl.includes(x))) amountCol = k;
  }
  if (!sectorCol || !amountCol) return null;
  let total = 0; let psl = 0;
  for (const row of data) {
    const amt = parseFloat(String(row[amountCol!] ?? 0)) || 0;
    const sec = String(row[sectorCol!] ?? '').trim().toLowerCase();
    total += amt;
    if ([...PSL_SECTORS].some((s) => sec.includes(s))) psl += amt;
  }
  if (total === 0) return null;
  const ratio = (psl / total) * 100;
  if (ratio >= 40) return { rule: 'PSL_RATIO', status: 'compliant', message: `PSL ratio is ${ratio.toFixed(1)}%, meeting the RBI mandate of 40%.` };
  const shortfall = ((40 / 100) - (psl / total)) * total;
  return {
    rule: 'PSL_RATIO', status: 'warning',
    message: `PSL ratio is ${ratio.toFixed(1)}%, below RBI mandate of 40%. Shortfall: ${(40 - ratio).toFixed(1)} pp (approx ₹${(shortfall / 1e7).toFixed(1)} Cr).`,
    psl_ratio: Math.round(ratio * 100) / 100, target: 40,
  };
}

function checkPmlaThreshold(data: ParsedRow[]): Record<string, any> | null {
  if (!data.length) return null;
  const THRESHOLD = 1_000_000;
  const keys = Object.keys(data[0]);
  let amountCol: string | null = null;
  let modeCol: string | null = null;
  for (const k of keys) {
    const kl = k.toLowerCase();
    if (!amountCol && ['amount','value','txn_amount'].some((x) => kl.includes(x))) amountCol = k;
    if (!modeCol && ['mode','channel','type','method'].some((x) => kl.includes(x))) modeCol = k;
  }
  if (!amountCol) return null;
  const triggers = data.filter((row) => {
    const amt = parseFloat(String(row[amountCol!] ?? 0)) || 0;
    const mode = modeCol ? String(row[modeCol] ?? '').toLowerCase() : '';
    return amt >= THRESHOLD && (mode.includes('cash') || mode === '');
  });
  if (!triggers.length) return null;
  return {
    rule: 'PMLA_CTR_THRESHOLD', status: 'warning',
    message: `${triggers.length} transaction(s) meet the PMLA CTR threshold of ₹10 lakh. CTRs must be filed with FIU-IND by the 15th of the following month.`,
    affected_count: triggers.length,
  };
}

export function runAllApplicableRules(question: string, data: ParsedRow[]): Record<string, any>[] {
  const annotations: Record<string, any>[] = [];
  const q = question.toLowerCase();

  if (['npa','asset class','classification','overdue','past due','dpd','sma','standard'].some((kw) => q.includes(kw))) {
    const r = checkNpaClassification(data);
    if (r) annotations.push(r);
  }
  if (['psl','priority sector','agriculture','msme','education','housing','weaker'].some((kw) => q.includes(kw))) {
    const r = checkPslRatio(data);
    if (r) annotations.push(r);
  }
  if (['cash','transaction','pmla','aml','suspicious','structuring','10 lakh','threshold'].some((kw) => q.includes(kw))) {
    const r = checkPmlaThreshold(data);
    if (r) annotations.push(r);
  }

  return annotations;
}
